radicle/src/storage/git/transport/local.rs

```diff
@@ -1,2 +1,107 @@
+//! The local git transport protocol.
 pub mod url;
 pub use url::{Url, UrlError};
+
+use std::cell::RefCell;
+use std::process;
+use std::str::FromStr;
+use std::sync::Once;
+
+use once_cell::sync::OnceCell;
+
+use crate::storage;
+use crate::storage::git::Storage;
+
+use super::ChildStream;
+
+thread_local! {
+    /// Stores a storage instance per thread.
+    static THREAD_STORAGE: OnceCell<Storage> = OnceCell::default();
+}
+
+/// Local git transport over the filesystem.
+#[derive(Default)]
+struct Local {
+    /// The child process we spawn.
+    child: RefCell<Option<process::Child>>,
+}
+
+impl git2::transport::SmartSubtransport for Local {
+    fn action(
+        &self,
+        url: &str,
+        service: git2::transport::Service,
+    ) -> Result<Box<dyn git2::transport::SmartSubtransportStream>, git2::Error> {
+        let url = Url::from_str(url).map_err(|e| git2::Error::from_str(e.to_string().as_str()))?;
+        let service: &str = match service {
+            git2::transport::Service::UploadPack | git2::transport::Service::UploadPackLs => {
+                "upload-pack"
+            }
+            git2::transport::Service::ReceivePack | git2::transport::Service::ReceivePackLs => {
+                "receive-pack"
+            }
+        };
+        let git_dir = THREAD_STORAGE
+            .with(|t| {
+                t.get()
+                    .map(|s| storage::git::paths::repository(&s, &url.repo))
+            })
+            .ok_or_else(|| git2::Error::from_str("local transport storage was not registered"))?;
+
+        let mut cmd = process::Command::new("git");
+
+        if let Some(ns) = url.namespace {
+            cmd.env("GIT_NAMESPACE", ns.to_string());
+        }
+
+        let mut child = cmd
+            .arg(service)
+            .arg(&git_dir)
+            .stdin(process::Stdio::piped())
+            .stdout(process::Stdio::piped())
+            .stderr(process::Stdio::inherit())
+            .spawn()
+            .map_err(|e| git2::Error::from_str(e.to_string().as_str()))?;
+
+        let stdin = child.stdin.take().expect("taking stdin is safe");
+        let stdout = child.stdout.take().expect("taking stdout is safe");
+
+        self.child.replace(Some(child));
+
+        Ok(Box::new(ChildStream { stdout, stdin }))
+    }
+
+    fn close(&self) -> Result<(), git2::Error> {
+        if let Some(mut child) = self.child.take() {
+            let result = child
+                .wait()
+                .map_err(|e| git2::Error::from_str(e.to_string().as_str()))?;
+
+            if !result.success() {
+                if let Some(code) = result.code() {
+                    log::error!("Child exited with error {}", code);
+                } else {
+                    log::error!("Child exited with unknown error code");
+                }
```

Comment:
should a failure be returned here instead?  The caller may be the only one with context to understand how its handled.

Reply:
Yeah good point, I only returned an error in the case the wait fails.